tools/communication/src/spi.cpp

```diff
@@ -0,0 +1,288 @@
+		 uint8_t bytes_number)
+{
+	auto *m2KSpiDesc = (m2k_spi_desc *) desc->extra;
+	auto samplesPerBit = (unsigned int) (m2KSpiDesc->sample_rate / desc->max_speed_hz);
+
+	//set the trigger on CS
+	libm2k::analog::M2kHardwareTrigger *trigger = m2KSpiDesc->digital->getTrigger();
+	trigger->setDigitalCondition(desc->chip_select, libm2k::analog::FALLING_EDGE);
+
+	//capture samples
+	std::vector<unsigned short> samples = m2KSpiDesc->digital->getSamples((bytes_number + 1) * samplesPerBit * 8);
+
+	//process samples
+	processSamples(desc, data, bytes_number, samples);
+}
+
+int32_t spi_init(struct spi_desc **desc,
+		 const struct spi_init_param *param)
+{
+	try {
+		auto spiDesc = new spi_desc;
+		auto m2KSpiDesc = new m2k_spi_desc;
+		m2k_spi_init *m2KSpiInit;
+
+		// initialize the attributes
+		unsigned int sampleRate = getValidSampleRate(param->max_speed_hz, SAMPLES_PER_CYCLE);
+		spiDesc->max_speed_hz = sampleRate / SAMPLES_PER_CYCLE;
+		spiDesc->mode = param->mode;
+		spiDesc->chip_select = param->chip_select;
+
+		m2KSpiInit = (m2k_spi_init *) param->extra;
+
+		m2KSpiDesc->clock = m2KSpiInit->clock;
+		m2KSpiDesc->mosi = m2KSpiInit->mosi;
+		m2KSpiDesc->miso = m2KSpiInit->miso;
+		m2KSpiDesc->bit_numbering = m2KSpiInit->bit_numbering;
+		m2KSpiDesc->context = m2KSpiInit->context;
+		m2KSpiDesc->digital = m2KSpiDesc->context->getDigital();
+		m2KSpiDesc->sample_rate = sampleRate;
+
+		m2KSpiDesc->digital->flushBufferIn();
+		m2KSpiDesc->digital->setKernelBuffersCountIn(1);
+
+		//set sampling frequencies
+		m2KSpiDesc->digital->setSampleRateOut(m2KSpiDesc->sample_rate);
+		m2KSpiDesc->digital->setSampleRateIn(m2KSpiDesc->sample_rate);
+
+		//enable the channels
+		setOutputChannel(spiDesc->chip_select, m2KSpiDesc->digital);
+		setOutputChannel(m2KSpiDesc->clock, m2KSpiDesc->digital);
+		setOutputChannel(m2KSpiDesc->mosi, m2KSpiDesc->digital);
+		setInputChannel(m2KSpiDesc->miso, m2KSpiDesc->digital);
+
+		m2KSpiDesc->digital->setOutputMode(spiDesc->chip_select, libm2k::digital::DIO_PUSHPULL);
+		m2KSpiDesc->digital->setOutputMode(m2KSpiDesc->clock, libm2k::digital::DIO_PUSHPULL);
+		m2KSpiDesc->digital->setOutputMode(m2KSpiDesc->mosi, libm2k::digital::DIO_PUSHPULL);
+		m2KSpiDesc->digital->setOutputMode(m2KSpiDesc->miso, libm2k::digital::DIO_PUSHPULL);
+
+		m2KSpiDesc->digital->setCyclic(false);
+
+		libm2k::analog::M2kHardwareTrigger *trigger = m2KSpiDesc->digital->getTrigger();
+		trigger->setDigitalCondition(m2KSpiDesc->clock, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(m2KSpiDesc->mosi, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(m2KSpiDesc->miso, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(spiDesc->chip_select, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalMode(libm2k::digital::DIO_OR);
+		trigger->setDigitalDelay(0);
+
+		//set the state of CS and CLK to idle
+		m2KSpiDesc->digital->setValueRaw(spiDesc->chip_select, libm2k::digital::HIGH);
+		if (spiDesc->mode >> 1u) {
```

Comment:
spiDesc->mode & SPI_CPOL would be more descriptive

Reply:
Modified in commit cd7956b.